File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {

    public void solve(int index,StringBuilder temp,String digits,HashMap<Character,String>map,ArrayList<String>result){
        if(index>=digits.length()){
            result.add(temp.toString());
            return;
        }
        char ch=digits.charAt(index);
        String letter=map.get(ch);

        for(int i=0;i<letter.length();i++){
            temp.append(letter.charAt(i));
            solve(index+1,temp,digits,map,result);
            temp.deleteCharAt(temp.length()-1);
        }

    }
    public List<String> letterCombinations(String digits) {
        ArrayList<String>result=new ArrayList<>();
        StringBuilder temp= new StringBuilder();
        if(digits.length()==0){
            return result;
        }
        HashMap<Character,String>map=new HashMap<>();
        map.put('2',"abc");
        map.put('3',"def");
        map.put('4',"ghi");
        map.put('5',"jkl");
        map.put('6',"mno");
        map.put('7',"pqrs");
        map.put('8',"tuv");
        map.put('9',"wxyz");
        solve(0,temp,digits,map,result);
        return result;
        
    }
}

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:

void solve(int index,string temp,string digits,unordered_map<char,string>&mp,vector<string>&result){
    if(index>=digits.size()){
        result.push_back(temp);
        return;
    }
    char ch=digits[index];
    string traverse=mp[ch];
    for(int i=0;i<traverse.size();i++){
        temp.push_back(traverse[i]);
        solve(index+1,temp,digits,mp,result);
        temp.pop_back();
    }
}

    vector<string> letterCombinations(string digits) {
        if(digits.length()==0){
            return {};
        }
        unordered_map<char,string>mp;
        string temp="";
        mp['2']="abc";
        mp['3']="def";
        mp['4']="ghi";
        mp['5']="jkl";
        mp['6']="mno";
        mp['7']="pqrs";
        mp['8']="tuv";
        mp['9']="wxyz";
        vector<string>result;
        solve(0,temp,digits,mp,result);
        return result;

        
    }
};